File: ios/StillMoment/Presentation/Views/GuidedMeditations/TrimEditor/TrimMarkHandle.swift
//
//  TrimMarkHandle.swift
//  Still Moment
//
//  Presentation Layer — purely visual trim mark (cut edge + grip knob) (shared-107).
//

import SwiftUI

/// A single trim mark (start or end): a thin full-height cut edge plus a grip knob
/// sitting clearly in the lower half of the waveform.
///
/// Purely visual — it never participates in hit testing. Touches are resolved
/// geometrically by the single track gesture in `TrimWaveformSection`, so overlapping
/// marks can never steal each other's touch. The active mark is wider, carries a glow
/// ring and a slow pulse (paused while dragging and under Reduce Motion). Exposed as
/// an adjustable accessibility element (±1 s nudges).
struct TrimMarkHandle: View {
    // MARK: Internal

    let time: TimeInterval
    let isActive: Bool
    let isDragging: Bool
    let trackWidth: CGFloat
    let duration: TimeInterval
    let onNudge: (TimeInterval) -> Void
    let accessibilityLabelText: String

    var body: some View {
        ZStack(alignment: .top) {
            self.bubble
            self.cutEdge
            self.knob
        }
        .frame(width: Self.layoutWidth, height: TrimWaveformView.height + Self.verticalOverhang * 2)
        .offset(x: self.markX - Self.layoutWidth / 2, y: -Self.verticalOverhang)
        .accessibilityElement()
        .accessibilityLabel(Text(self.accessibilityLabelText))
        .accessibilityValue(Text(EditSheetState.formatTime(self.time)))
        .accessibilityAdjustableAction { direction in
            switch direction {
            case .increment: self.onNudge(1)
            case .decrement: self.onNudge(-1)
            @unknown default: break
            }
        }
        .onChange(of: self.isDragging) { dragging in
            self.updatePulse(dragging: dragging)
        }
        .onChange(of: self.isActive) { _ in
            self.updatePulse(dragging: self.isDragging)
        }
        .onAppear { self.updatePulse(dragging: false) }
    }

    // MARK: Private

    @Environment(\.themeColors)
    private var theme

    @Environment(\.accessibilityReduceMotion)
    private var reduceMotion

    @State private var pulsing = false

    private static let layoutWidth: CGFloat = 44
    /// How far the cut edge extends beyond the waveform top/bottom.
    private static let verticalOverhang: CGFloat = 12
    /// Vertical center of the knob, as fraction of the waveform height — clearly
    /// in the lower (mark) zone so the visual matches the touch split.
    private static let knobCenterFraction: CGFloat = 0.74

    private var markX: CGFloat {
        TrimGeometry.x(for: self.time, duration: self.duration, width: self.trackWidth)
    }

    private var markGradient: LinearGradient {
        LinearGradient(
            colors: [self.theme.playGradientTop, self.theme.playGradientBot],
            startPoint: .top,
            endPoint: .bottom
        )
    }

    private var cutEdge: some View {
        RoundedRectangle(cornerRadius: 3)
            .fill(self.markGradient)
            .frame(width: self.isActive ? 4 : 3)
            .opacity(self.isActive ? 1 : 0.7)
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .animation(.easeOut(duration: 0.12), value: self.isActive)
    }

    private var knob: some View {
        RoundedRectangle(cornerRadius: 8)
            .fill(self.markGradient)
            .frame(width: self.isActive ? 20 : 16, height: self.isActive ? 44 : 38)
            .overlay(
                HStack(spacing: 3) {
                    self.grip
                    self.grip
                }
            )
            .overlay(
                RoundedRectangle(cornerRadius: 8)
                    .strokeBorder(self.theme.interactive.opacity(self.isActive ? 0.35 : 0), lineWidth: 2)
                    .padding(-2)
            )
            .shadow(color: self.theme.cardShadow, radius: 4, y: 2)
            .scaleEffect(self.pulsing ? 1.1 : 1)
            .animation(.easeOut(duration: 0.12), value: self.isActive)
            .position(
                x: Self.layoutWidth / 2,
                y: Self.verticalOverhang + TrimWaveformView.height * Self.knobCenterFraction
            )
    }

    private var grip: some View {
        Capsule()
            .fill(self.theme.textOnInteractive.opacity(0.5))
            .frame(width: 2, height: 13)
    }

    @ViewBuilder private var bubble: some View {
        if self.isDragging {
            TrimTimeBubble(
                time: self.time,
                anchorX: self.markX,
                trackWidth: self.trackWidth,
                background: self.theme.interactive,
                textColor: self.theme.textOnInteractive
            )
            .offset(y: -30)
        }
    }

    private func updatePulse(dragging: Bool) {
        let shouldPulse = self.isActive && !dragging && !self.reduceMotion
        guard shouldPulse else {
            self.pulsing = false
            return
        }
        withAnimation(.easeOut(duration: 1.8).repeatForever(autoreverses: true)) {
            self.pulsing = true
        }
    }
}

File: android/app/src/main/kotlin/com/stillmoment/presentation/ui/meditations/trimeditor/TrimMarkHandle.kt
package com.stillmoment.presentation.ui.meditations.trimeditor

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.stillmoment.presentation.ui.theme.LocalStillMomentColors

/**
 * A single trim mark (start or end): a thin full-height cut edge plus a grip knob sitting
 * clearly in the lower half of the waveform (shared-107).
 *
 * Purely visual — it never participates in hit testing. Touches are resolved geometrically
 * by the single track gesture in [TrimWaveformSection] ([TrimHitTesting]), so overlapping
 * marks can never steal each other's touch. The active mark is wider and carries a glow ring.
 *
 * 1:1 port of iOS `TrimMarkHandle` (without the pulse animation — kept static for simplicity;
 * the active mark is still emphasised by width + glow). Drawn as a full-size overlay over the
 * waveform; positions itself at [timeMs] inside [window].
 */
@Composable
fun TrimMarkHandle(
    timeMs: Long,
    isActive: Boolean,
    trackWidthPx: Float,
    window: ClosedRange<Long>,
    waveformHeightPx: Float,
    modifier: Modifier = Modifier
) {
    val theme = LocalStillMomentColors.current
    val gradient = Brush.verticalGradient(listOf(theme.playGradientTop, theme.playGradientBot))
    Box(modifier = modifier.fillMaxSize()) {
        Canvas(modifier = Modifier.fillMaxSize()) {
            if (trackWidthPx <= 0f) {
                return@Canvas
            }
            val markX = TrimGeometry.x(timeMs, window, trackWidthPx)
            drawCutEdge(markX, isActive, gradient)
            drawKnob(markX, waveformHeightPx, isActive, gradient, theme.interactive, theme.textOnInteractive)
        }
    }
}

private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawCutEdge(
    markX: Float,
    isActive: Boolean,
    gradient: Brush
) {
    val edgeWidth = (if (isActive) 4f else 3f).dp.toPx()
    val corner = CornerRadius(3.dp.toPx(), 3.dp.toPx())
    drawRoundRect(
        brush = gradient,
        topLeft = Offset(markX - edgeWidth / 2f, 0f),
        size = Size(edgeWidth, size.height),
        cornerRadius = corner,
        alpha = if (isActive) 1f else 0.7f
    )
}

private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawKnob(
    markX: Float,
    waveformHeightPx: Float,
    isActive: Boolean,
    gradient: Brush,
    glowColor: Color,
    gripColor: Color
) {
    val knobWidth = (if (isActive) 20f else 16f).dp.toPx()
    val knobHeight = (if (isActive) 44f else 38f).dp.toPx()
    val knobCenterY = waveformHeightPx * KNOB_CENTER_FRACTION
    val topLeft = Offset(markX - knobWidth / 2f, knobCenterY - knobHeight / 2f)
    val corner = CornerRadius(8.dp.toPx(), 8.dp.toPx())

    if (isActive) {
        drawRoundRect(
            color = glowColor.copy(alpha = 0.35f),
            topLeft = Offset(topLeft.x - 2.dp.toPx(), topLeft.y - 2.dp.toPx()),
            size = Size(knobWidth + 4.dp.toPx(), knobHeight + 4.dp.toPx()),
            cornerRadius = corner,
            style = androidx.compose.ui.graphics.drawscope.Stroke(width = 2.dp.toPx())
        )
    }
    drawRoundRect(brush = gradient, topLeft = topLeft, size = Size(knobWidth, knobHeight), cornerRadius = corner)

    // Two grip lines.
    val gripHeight = 13.dp.toPx()
    val gripWidth = 2.dp.toPx()
    val gripSpacing = 3.dp.toPx()
    val gripTop = knobCenterY - gripHeight / 2f
    listOf(markX - gripSpacing, markX + gripSpacing).forEach { gx ->
        drawRoundRect(
            color = gripColor.copy(alpha = 0.5f),
            topLeft = Offset(gx - gripWidth / 2f, gripTop),
            size = Size(gripWidth, gripHeight),
            cornerRadius = CornerRadius(gripWidth, gripWidth)
        )
    }
}

/** Vertical center of the knob as a fraction of the waveform height — in the lower (mark) zone. */
private const val KNOB_CENTER_FRACTION = 0.74f
